### ISSUE TITLE
Running out of /tmp space leaves corrupt file on GCS

### ISSUE THREAD

**Issue by squee1945** [GoogleCloudPlatform/gcsfuse#77]
I am trying to work with gcsfuse using large files - so far, I'm up to a 10GB file, but hoping to get up to 200+ GB.

When I was trying to copy a 10 GB file from GCE to GCS via gcsfuse, I got the following error:

```
$ cp /data/files/ten-gig.txt bucket
fuse: 2015/06/16 14:26:08.539290 (LookUpInode(parent=1, name="ten-gig.txt")) error: no such file or directory
fuse: 2015/06/16 14:26:08.540425 (LookUpInode(parent=1, name="ten-gig.txt")) error: no such file or directory
fuse: 2015/06/16 14:28:21.853851 (WriteFile(inode=5)) error: write /tmp/gcsfuse851117821: no space left on device
cp: writing `bucket/ten-gig.txt': Input/output error
cp: failed to extend `bucket/ten-gig.txt': Input/output error
fuse: 2015/06/16 14:30:43.573473 (ReleaseFileHandle(inode=5)) error: function not implemented

# GCS (destination) files
$ ll bucket
-rw-r--r-- 1 jasonco jasonco 7.9G Jun 16 14:30 ten-gig.txt

# local (source) files
$ ll /data/files/
-rw-r--r-- 1 jasonco jasonco 9.8G Jun 15 20:59 ten-gig.txt
```

So it seems that my boot disk ran out of space on this operation.

1. Is there any way to avoid the extra tmp copy? These are large files, so the extra overhead concerns me. Or at least, can I specify an alternate work space?
2. More important than (1), it appears that the file on GCS is fine, except it's about 2 GB too small, so it's clearly corrupt. Can gcsfuse correctly clean up in this event?

**Comment by jacobsa** [comment]
Thanks for the report.

1. No, everything must bounce off of a local staging file before going out to GCS. This is because posix file system semantics let you do random writes but GCS doesn't, and to avoid a potentially very long-lived upload operation to GCS that would likely fail. That said, you can control the temporary directory and how much space it will use for "clean" files with `--temp-dir` and `--temp-dir-bytes`. See also [here](18767951ea35769e7740a9c741b31884e8098e41).

2. This is an interesting question. The user software (in this case cp) saw a write error, so the file system is not guaranteeing that the data made it though okay. But it's forced to close the file either way, which causes an object to be written to GCS unconditionally. I would guess that cp returned a non-zero exit code, so e.g. a shell script running it would fail correctly. Really, this matches what would happen if there were a problem writing to the local disk (i.e. you would see a truncated file), so I'm not sure that gcsfuse is doing anything wrong here.

**Comment by squee1945** [comment]
Could the process copy the bytes up to GCS over time in a temp file, and then as the final step, use `gsutil mv` to put them on the correct path? A failed/corrupt file would stick around, but at least it would be on a different filename.

**Comment by jacobsa** [comment]
Assuming `gsutil mv` works by composing objects, gcsfuse actually already does this when you've only appended to an object since the last flush. So if you call fsync(2) regularly, this is what will happen, with two caveats:

1. It can only be done if the object isn't already composed of 1024 components. (This is a GCS limit.)
2. It is only done when the object is already of a certain minimum length.

Of course that doesn't help with the particular case of `cp`, which doesn't call fsync(2). Still, I believe the behavior here matches a real file system, so I don't think there is a bug. I will test with an out of space error on a real file system to confirm.

**Comment by jacobsa** [comment]
Sorry, my last comment was misleading: gcsfuse does what I describe, but it doesn't match what you describe because it's composing over the final location each time.

**Comment by squee1945** [comment]
I'm confused now. I ended up with a file on GCS that was 2GB too small, but under the correct filename. So a client on a different system would have no way to know that the file on GCS was corrupt.

**Comment by jacobsa** [comment]
Right, and this matches what happens on a normal file system:

```
% cd /tmp

% du -h bar
11G bar

% cp bar foo; echo $?
cp: error writing ‘foo’: No space left on device
cp: failed to extend ‘foo’: No space left on device
1

% du -h foo bar
8.1G foo
11G bar
```

**Comment by squee1945** [comment]
But that's different, you're running out of space on the destination file system. In my case, the destination file system is GCS, which presumably is not running out of space.

The thing that's running out of space in my case is an internal implementation detail of gcsfuse.

**Comment by jacobsa** [comment]
The point isn't running out of space, it's that the file system experienced an error during write(2). My example would be the same for any error; it's just that running out of space is easy to simulate.

gcsfuse is simulating a posix file system, and when `cp` sees an error during a write to a posix file system it closes the file, leaving behind something shorter than what you intended, and exits with a non-zero status code. It doesn't matter whether this is gcsfuse, ext2, or whatever file system. This seems to be working as intended.

**Comment by jacobsa** [comment]
Elaborating to make sure we're on the same page. I'm assuming that you're copying a source file `src` to a destination file `dst`, and that your goals are:

* If `dst` doesn't already exist, an observer will only ever see it not exist or will see it exist with exactly the contents of `src`.
* If `dst` already exists, an observer will only ever see its old contents or exactly the contents of `src`.

I'm also assuming that `cp` works as follows:

1. Call open(2) with `O_CREAT|O_TRUNC` for `dst`.
2. Open `src` and repeatedly use read(2) and write(2) to read data from `src` and write it to `dst`.
3. Call close(2) for `dst`.

Then doing `cp src dst` doesn't fulfill your goals on any file system. Here are some ways this goes wrong:

* `dst` is visible as an empty file as soon as it is opened.
* `dst` is visible while some contents have been written but not all.
* The system might lose power, or the `cp` process crash.
* There may be an out of space error, or some other type of error.

The typical way you accomplish this using a posix file system is instead:

    cp src tmp && mv tmp dst

This works because:

*   In the case where `dst` doesn't already exist, the directory entry is not created until the inode contains the desired data.

    (I'm assuming here that posix respects causality: if you write(2) before you rename(2), then the data is visible. I can't find this spelled out anywhere, but it seems reasonable to assume.)

*   In the case where `dst` already exists, posix [guarantees](http://pubs.opengroup.org/onlinepubs/9699919799/functions/rename.html) that the change in what it points to is atomic:

    > If the link named by the _new_ argument exists, it shall be removed and _old_ renamed to _new_. In this case, a link named _new_ shall remain visible to other processes throughout the renaming operation and refer either to the file referred to by _new_ or _old_ before the operation began.

Unfortunately, this won't work on gcsfuse because it [doesn't support renaming](https://github.com/googlecloudplatform/gcsfuse/blob/db7997eda35828e80d23892c38a8bec14d4bf751/docs/semantics.md#missing-features). However the reason it doesn't support renaming is that I thought posix required the entire operation to be atomic, rather than just the view of _new_. In other words, that `tmp` must not exist at the same time as `dst` when the latter is being newly created. But the man page for rename(2) on Linux explicitly states this is not the case:

> However, when overwriting there will probably be a window in which both oldpath and newpath refer to the file being renamed.

So in other words I can probably support `mv`, at least for regular files. Would this be good enough for you?

**Comment by squee1945** [comment]
Without getting into long discussions about the state of filesystems today (though I appreciate your message here because it does point out flaws in the current state of the art), I'm sure you can understand that someone would welcome any improvement in the overall reliability of a filesystem. E.g., any percentage increase in the reliability of a cp command helps alleviate the amount of mitigating code that needs to be written by the client.

In the case of gcsfuse, it seems the extra tmp copy adds another point of fragility in the overall process (and thus this issue), so anything that can be done to minimize it will only help the overall adoption of the tool.

I'm not sure if you're suggesting that it would be left up to the client to first `cp` to a temp file and then `mv` it, or if gcsfuse would do this automatically to mitigate the extra temp copy. If the former, in my particular case, it probably doesn't help because the tools that are in use by my customer are black box and they're not able to change that logic.

**Comment by jacobsa** [comment]
I'm suggesting the former: when interacting with a posix file system, if you want the guarantees I mentioned, it's not safe to just use `cp`. gcsfuse's reason for existing is to look like a posix file system, and any magic behavior intended to improve on the posix semantics is out of scope.

**Comment by jacobsa** [comment]
(Closed because I don't see a concrete proposal for something gcsfuse can do to help here.)

**Comment by jacobsa** [comment]
Sorry, I wrote the last two comments one-handed while holding a sleeping baby, and was more terse than I should have been. Let me try again.

What I failed to explain is that when you do `cp src dst`, the kernel doesn't send gcsfuse a single request. Actually it's a bunch of requests, something like:

* Look up `src`, return an inode number.
* Open the `src` inode for reading.
* Look up `dst`, confirm it doesn't exist.
* Create and open a new zero-length file named `dst`, return the inode number.
* Read 128 KiB from the `src` inode.
* Write 128 KiB to the `dst` inode.
* Read 128 KiB from the `src` inode.
* Write 128 KiB to the `dst` inode.
* ...
* Flush and close the `dst` inode.

There's no way that gcsfuse can tell that this resulted from a simple `cp src dst` from the user, so it can't divine that it should be all or nothing—instead it must serve each write as it comes in individually.

I've closed the issue because I can't think of any way that gcsfuse can help here. Do let me know if you've got any ideas.
